runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+				SlashingOffenceKind::ForInvalid => {
+					let is_empty = <PendingForInvalidLosers<T>>::try_mutate_exists(
+						&session_index,
+						&candidate_hash,
+						try_remove,
+					)?;
+
+					let winners = if is_empty {
+						<ForInvalidWinners<T>>::take(&session_index, &candidate_hash)
+							.unwrap_or_default()
+					} else {
+						<ForInvalidWinners<T>>::get(&session_index, &candidate_hash)
+							.unwrap_or_default()
+					};
+
+					let offence = ForInvalidOffence::new(
+						session_index,
+						candidate_hash,
+						validator_set_count,
+						offender,
+					);
+
+					// We can't really submit a duplicate report
+					// unless there's a bug.
+					<T::HandleReports as HandleReports<T>>::report_for_invalid_offence(
+						winners, offence,
+					)
+					.map_err(|_| Error::<T>::DuplicateSlashingReport)?;
+				},
+				SlashingOffenceKind::AgainstValid => {
+					let is_empty = <PendingAgainstValidLosers<T>>::try_mutate_exists(
+						&session_index,
+						&candidate_hash,
+						try_remove,
+					)?;
+
+					let winners = if is_empty {
+						<AgainstValidWinners<T>>::take(&session_index, &candidate_hash)
+							.unwrap_or_default()
+					} else {
+						<AgainstValidWinners<T>>::get(&session_index, &candidate_hash)
+							.unwrap_or_default()
+					};
+
+					// submit an offence report
+					let offence = AgainstValidOffence::new(
+						session_index,
+						candidate_hash,
+						validator_set_count,
+						offender,
+					);
+					// We can't really submit a duplicate report
+					// unless there's a bug.
+					<T::HandleReports as HandleReports<T>>::report_against_valid_offence(
+						winners, offence,
+					)
+					.map_err(|_| Error::<T>::DuplicateSlashingReport)?;
+				},
+			}
+
+			Ok(Pays::No.into())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			Self::validate_unsigned(source, call)
+		}
+
+		fn pre_dispatch(call: &Self::Call) -> Result<(), TransactionValidityError> {
+			Self::pre_dispatch(call)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Called by the initializer to initialize the disputes slashing module.
+	fn initializer_initialize(_now: T::BlockNumber) -> Weight {
+		0
+	}
+
+	/// Called by the initializer to finalize the disputes slashing pallet.
+	fn initializer_finalize() {}
+
+	/// Called by the initializer to note a new session in the disputes slashing pallet.
+	fn initializer_on_new_session(
+		session_index: SessionIndex,
+		validator_set_count: ValidatorSetCount,
+	) {
+		<ValidatorSetCounts<T>>::insert(session_index, validator_set_count);
+
+		let config = <crate::configuration::Pallet<T>>::config();
+		if session_index <= config.dispute_period + 1 {
+			return
+		}
+
+		let old_session = session_index - config.dispute_period - 1;
+
+		// This should be small, as disputes are rare, so no limit is fine.
```

Comment:
how much of an effort is it to make it not be unbounded? If a lot of disputed happen due to a bug, will the chian survive?

Reply:
Don't know, but we'll also need to fix many other places, e.g.
https://github.com/paritytech/polkadot/blob/e7461a6bd020250fed010ba56b78b854ad463064/runtime/parachains/src/disputes.rs#L866-L886

Reply:
what I am saying is coming from a pain point that the slashing code was written with a similar attitude of being unbounded (and at the time we actually didn't have good weights), and it has been 2+ years and that code is still a point of issue and worry for us. I am concerned that we are repeating the same mistake. 

What I am saying is, perhaps the effort of making this code is bounded now is `x`, but the effort of doing it 2 years from now once you all have forgotten about it, and a million other code is build on top of it, is `10x`, so it is worthwhile doing it now right form the get go.

That being said, if this code is only ever going to live on the relay chain, it can generally be lenient when it comes to weight.

Reply:
Yes, this is relay chain only code and we have spam slot mechanism to bound the number of disputes.

I do agree that the technical dept costs tend to grow exponentially with time. But arguably this PR doesn't create a new unbounded removal, it was there in the disputes before, so can be addressed separately from this PR.